Hw 02 solution

kopervas opened the pull request.

kopervas committed:
Revert "Revert "1.0.0 First commit.""

This reverts commit 4871f167f807e6bbc0caa61e8b00c0b945f2ec05.

kopervas committed:
1.0.1 Secont commit - check ci.yml.

kopervas committed:
1.0.2 Third commit - created the first model Book, BookRepository and BookService.

kopervas committed:
1.0.3 Fourth commit. Bug fixes according to comments from Liliia Ponomarenko.

kopervas committed:
1.0.4 Fifth commit. Bug fixes according to comments from Olena Bruiako.

kopervas committed:
1.0.5 Sixth commit. Bug fixes according to comments from Olena Bruiako.

kopervas committed:
1.0.6 Seventh commit. Added Controller, DTO, implemented MapStruct.

kopervas committed:
1.0.7 Revert commit. Undo all changes made to the code in commit 1.0.6..

kopervas committed:
Spring Boot Intro (#3)

* Revert "Revert "1.0.0 First commit.""

This reverts commit 4871f167f807e6bbc0caa61e8b00c0b945f2ec05.

* 1.0.1 Secont commit - check ci.yml.

* 1.0.2 Third commit - created the first model Book, BookRepository and BookService.

* 1.0.3 Fourth commit. Bug fixes according to comments from Liliia Ponomarenko.

* 1.0.4 Fifth commit. Bug fixes according to comments from Olena Bruiako.

* 1.0.5 Sixth commit. Bug fixes according to comments from Olena Bruiako.

* 1.0.6 Seventh commit. Added Controller, DTO, implemented MapStruct.

* 1.0.7 Revert commit. Undo all changes made to the code in commit 1.0.6..

kopervas committed:
Revert "1.0.7 Revert commit. Undo all changes made to the code in commit 1.0.6.."

This reverts commit 3eef526d0e0f9e713e93c4409651cec40258587c.

kopervas committed:
Revert "1.0.7 Revert commit. Undo all changes made to the code in commit 1.0.6.."

This reverts commit 3eef526d0e0f9e713e93c4409651cec40258587c.

kopervas committed:
Merge remote-tracking branch 'origin/hw-02-solution' into hw-02-solution

kopervas committed:
1.0.8 Correction of errors according to comments from Olena Bruiako.

kopervas committed:
1.0.9 Correction of errors according to comments from Liliia Ponomarenko.

kopervas committed:
1.1.0 Correction of errors according to comments from Olena Bruiako.

kopervas committed:
1.2.0 Release of the task to the topic 'Spring Data JPA'. Used Lombok, Liquibase, implemented soft delete concept for the main entity.

kopervas committed:
1.2.0 Release of the task to the topic 'Spring Data JPA'. Used Lombok, Liquibase, implemented soft delete concept for the main entity.

kopervas committed:
1.2.1 Release with bug fixes according to Olena Bruiako's comments.

Elena-Bruyako commented on liquibase.properties:
```diff
@@ -0,0 +1,5 @@
+changeLogFile=src/main/resources/db/changelog/db.changelog-master.yaml
```
replace to `resources` folder

Elena-Bruyako commented on src/main/java/com/bookstore/controller/BookController.java:
```diff
@@ -0,0 +1,46 @@
+
+@RequiredArgsConstructor
+@RestController
+@ResponseStatus(HttpStatus.CREATED)
```
```suggestion
```

Elena-Bruyako commented on src/main/java/com/bookstore/controller/BookController.java:
```diff
@@ -0,0 +1,46 @@
+        return bookService.findById(id);
+    }
+
+    @PostMapping
```
```suggestion
    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
```

Elena-Bruyako commented on src/main/java/com/bookstore/service/impl/BookServiceImpl.java:
```diff
@@ -11,14 +16,35 @@
+    @Override
+    public BookDto update(Long id, UpdateBookRequestDto requestDto) {
+        Book book = bookRepository.findById(id)
+                .orElseThrow(() -> new RuntimeException("Book not found with id: " + id));
```
```suggestion
                .orElseThrow(() -> new EntityNotFoundException("Book not found with id: " + id));
```

Elena-Bruyako reviewed (changes requested):
see comments

kopervas committed:
1.2.2 Second release with bug fixes according to Olena Bruiako's comments.

Elena-Bruyako commented on liquibase.properties:
```diff
@@ -0,0 +1,5 @@
+changeLogFile=changeLogFile=db/changelog/db.changelog-master.yaml
```
this file should be in `resources` folder

Elena-Bruyako commented on src/test/resources/application.properties:
```diff
@@ -3,5 +3,6 @@ spring.datasource.driverClassName=org.h2.Driver
 spring.datasource.password=password
 spring.jpa.database-platform=org.hibernate.dialect.H2Dialect
-spring.jpa.hibernate.ddl-auto=create-drop
+spring.jpa.hibernate.ddl-auto=none
```
```suggestion
```

Elena-Bruyako replied:
```suggestion
spring.jpa.hibernate.ddl-auto=validate
```

Elena-Bruyako reviewed (changes requested):
see comments

kopervas committed:
1.2.3 Third release with bug fixes according to Olena Bruiako's comments.

Elena-Bruyako reviewed (changes requested):
one comment

kopervas committed:
1.2.4 Release with bug fixes according to Olena Bruiako's comments.
 Was:
   spring.jpa.hibernate.ddl-auto=none
 Became:
   spring.jpa.hibernate.ddl-auto=validate

kopervas committed:
1.3.0 An additional task to the chapter "Criteria Query in Spring Boot".

Elena-Bruyako commented on src/main/java/com/bookstore/controller/BookController.java:
```diff
@@ -0,0 +1,52 @@
+
+    @GetMapping("/search")
+    public List<BookDto> searchBooks(BookSearchParameters searchParameters) {
+        return null;
```
provide implementation

Elena-Bruyako commented on src/main/java/com/bookstore/repository/book/AuthorSpecificationProvider.java:
```diff
@@ -0,0 +1,20 @@
+public class AuthorSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "author";
```
create constant

Elena-Bruyako commented on src/main/java/com/bookstore/repository/book/IsbnSpecificationProvider.java:
```diff
@@ -0,0 +1,20 @@
+public class IsbnSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "isbn";
```
create constant

Elena-Bruyako commented on src/main/java/com/bookstore/repository/book/TitleSpecificationProvider.java:
```diff
@@ -0,0 +1,20 @@
+public class TitleSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "title";
```
create constant

Elena-Bruyako commented on src/main/java/com/bookstore/repository/impl/BookSpecificationBuilder.java:
```diff
@@ -0,0 +1,33 @@
+    public Specification<Book> build(BookSearchParameters searchParameters) {
+        Specification<Book> spec = Specification.where(null);
+        if (searchParameters.title() != null && searchParameters.title().length > 0) {
+            spec = spec.and(phoneSpecificationProviderManager.getSpecificationProvider("title")
```
create constants for literals

Elena-Bruyako commented on src/main/java/com/bookstore/service/impl/BookServiceImpl.java:
```diff
@@ -1,24 +1,61 @@
+    public List<BookDto> search(BookSearchParameters params) {
+        Specification<Book> bookSpecification = bookSpecificationBuilder.build(params);
+
+        return bookRepository.findAll(bookSpecification).stream().map(bookMapper::toDto).toList();
```
```suggestion
        return bookRepository.findAll(bookSpecification)
        .stream()
        .map(bookMapper::toDto)
        .toList();
```

Elena-Bruyako reviewed (changes requested):
There are changes from previous HW, fix this please
See other comments

kopervas committed:
1.3.1 Errors correction according to comments from Olena Bruiako from 29.08.2025.

kopervas committed:
1.4.0 Implementation of GlobalExceptionHandler and Data Validation.
